src/Microsoft.TestPlatform.CrossPlatEngine/Execution/BaseRunTests.cs

```diff
@@ -363,122 +384,174 @@ private TimeSpan RunTestsInternal()
         [SuppressMessage("Microsoft.Design", "CA1031:DoNotCatchGeneralExceptionTypes", Justification = "This methods must call all possible executors and not fail on crash in any executor.")]
         private bool RunTestInternalWithExecutors(IEnumerable<Tuple<Uri, string>> executorUriExtensionMap, long totalTests)
         {
-            double totalTimeTakenByAdapters = 0;
-
-            var executorsFromDeprecatedLocations = false;
-
-            // Call the executor for each group of tests.
-            var exceptionsHitDuringRunTests = false;
-
-            // Collecting Total Number of Adapters Discovered in Machine
+            // Collecting Total Number of Adapters Discovered in Machine.
             this.requestData.MetricsCollection.Add(TelemetryDataConstants.NumberOfAdapterDiscoveredDuringExecution, executorUriExtensionMap.Count());
 
+            var attachedToTestHost = false;
+            var executorCache = new Dictionary<string, LazyExtension<ITestExecutor, ITestExecutorCapabilities>>();
             foreach (var executorUriExtensionTuple in executorUriExtensionMap)
             {
-                // Get the executor
+                // Avoid processing the same executor twice.
+                if (executorCache.ContainsKey(executorUriExtensionTuple.Item1.AbsoluteUri))
+                {
+                    continue;
+                }
+
+                // Get the extension manager.
                 var extensionManager = this.GetExecutorExtensionManager(executorUriExtensionTuple.Item2);
 
                 // Look up the executor.
                 var executor = extensionManager.TryGetTestExtension(executorUriExtensionTuple.Item1);
-                if (executor != null)
+                if (executor == null)
                 {
-                    try
-                    {
-                        if (EqtTrace.IsVerboseEnabled)
-                        {
-                            EqtTrace.Verbose(
-                                "BaseRunTests.RunTestInternalWithExecutors: Running tests for {0}",
-                                executor.Metadata.ExtensionUri);
-                        }
-
-                        // set the active executor
-                        this.activeExecutor = executor.Value;
+                    // Commenting this out because of a compatibility issue with Microsoft.Dotnet.ProjectModel released on nuGet.org.
+                    // this.activeExecutor = null;
+                    // var runtimeVersion = string.Concat(PlatformServices.Default.Runtime.RuntimeType, " ",	
+                    // PlatformServices.Default.Runtime.RuntimeVersion);
```

Comment:
nit: can we clear?

Reply:
well, more likely that we'd want to fix. :) 

---
In reply to: [423927679](https://github.com/microsoft/vstest/pull/2325#discussion_r423927679) [](ancestors = 423927679)

Reply:
Tried clearing but I was asked to bring it back to keep the historic context.

As this turns up as a change only because of a re-write I did for the `RunTestInternalWithExecutors` method (so it's not related to the current feature) and we already had our fair share of compat issues with this change I'd rather not fix it here.

Maybe it's better to open a low prio issue for this alone and have it contained to reduce the risk of having other compat issues pop up here.